driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,195 @@
+
+        List<Document> results = aggregate(
+                match(expr(current().getInteger("x").eq(of(1)))));
+
+        assertEquals(
+                Arrays.asList(Document.parse("{_id: 3, x: 1, y: 3}")),
+                results);
+    }
+
+    @Test
+    public void currentAsMapMatchTest() {
+        col.insertMany(Arrays.asList(
+                Document.parse("{_id: 1, x: 0, y: 2}"),
+                Document.parse("{_id: 2, x: 0, y: 3}"),
+                Document.parse("{_id: 3, x: 1, y: 3}")));
+
+        List<Document> results = aggregate(
+                match(expr(Expressions.<NumberExpression>currentAsMap()
+                        .entrySet()
+                        .map(e -> e.getValue())
+                        .sum(v -> v).eq(of(7)))));
+
+        assertEquals(
+                Arrays.asList(Document.parse("{_id: 3, x: 1, y: 3}")),
+                results);
+    }
+
+    @Test
+    public void projectTest() {
+        col.insertMany(Arrays.asList(
+                Document.parse("{_id: 1, x: 0, y: 2}")));
+
+        List<Document> expected = Arrays.asList(Document.parse("{_id: 1, x: 0, c: 2}"));
+
+        // old, using "$y"
+        Bson projectOld = project(fields(include("x"), computed("c",
+                "$y")));
+        assertEquals("{'$project': {'x': 1, 'c': '$y'}}", bsonToString(projectOld));
+        assertEquals(expected,
+                aggregate(projectOld));
+
+        // new, using current() with add/subtract
+        Bson projectNew = project(fields(include("x"), computed("c",
+                current().getInteger("y").add(10).subtract(10))));
+        assertEquals(
+                "{'$project': {'x': 1, 'c': "
+                        + "{'$subtract': [{'$add': [{'$getField': "
+                        + "{'input': '$$CURRENT', 'field': 'y'}}, 10]}, 10]}}}",
+                bsonToString(projectNew));
+        assertEquals(expected,
+                aggregate(projectNew));
+    }
+
+    @Test
+    public void projectTest2() {
+        col.insertMany(Arrays.asList(Document.parse("{_id: 0, x: 1}")));
+
+        // new, nestedArray
+        Bson projectNestedArray = project(fields(excludeId(), computed("nestedArray", ofArray(
+                current().getInteger("x").max(of(4)),
+                current().getInteger("x"),
+                of(0), of(1), of(true), of(false)
+        ))));
+        assertEquals(
+                Arrays.asList(Document.parse("{ nestedArray: [ 4, 1, 0, 1, true, false ] }")),
+                aggregate(projectNestedArray));
+
+        // new, document
+        Bson projectDocument = project(fields(computed("nested",
+                // the below is roughly: "{ x: {$max : ['$x', 4] }}"
+                of(Document.parse("{x: 9}")).setField("x", current().getInteger("x").max(of(4)))
+        )));
+        assertEquals(
+                Arrays.asList(Document.parse("{_id: 0, nested: { x: 4 } }")),
+                aggregate(projectDocument));
```

Comment:
Shouldn't `nested.x` be set to the max of 9 and 4 here? Based on the assertion it is set to 4 for some reason.

Reply:
`of(Document.parse("{x: 9}"))` is not the current document, it is a literal. It is having its `x` value correctly set ("overwritten") to the maximum of 1 and 4.

Reply:
Thank you, I was reading this absolutely incorrectly.